contracts/0.4.24/nos/NodeOperatorsRegistry.sol

```diff
@@ -138,228 +158,260 @@ contract NodeOperatorsRegistry is INodeOperatorsRegistry, IsContract, AragonApp
+     */
+    function setNodeOperatorRewardAddress(uint256 _id, address _rewardAddress)
+        external
         authP(SET_NODE_OPERATOR_ADDRESS_ROLE, arr(_id, uint256(_rewardAddress)))
         operatorExists(_id)
-        validAddress(_rewardAddress)
+        onlyNonZeroAddress(_rewardAddress)
     {
         require(operators[_id].rewardAddress != _rewardAddress, "NODE_OPERATOR_ADDRESS_IS_THE_SAME");
         operators[_id].rewardAddress = _rewardAddress;
         emit NodeOperatorRewardAddressSet(_id, _rewardAddress);
     }
 
     /**
-      * @notice Set the maximum number of validators to stake for the node operator #`_id` to `_stakingLimit`
-      */
-    function setNodeOperatorStakingLimit(uint256 _id, uint64 _stakingLimit) external
+     * @notice Set the maximum number of validators to stake for the node operator #`_id` to `_stakingLimit`
+     */
+    function setNodeOperatorStakingLimit(uint256 _id, uint64 _stakingLimit)
+        external
         authP(SET_NODE_OPERATOR_LIMIT_ROLE, arr(_id, uint256(_stakingLimit)))
         operatorExists(_id)
     {
         require(operators[_id].stakingLimit != _stakingLimit, "NODE_OPERATOR_STAKING_LIMIT_IS_THE_SAME");
         _increaseKeysOpIndex();
-        operators[_id].stakingLimit = _stakingLimit;
-        emit NodeOperatorStakingLimitSet(_id, _stakingLimit);
+        _setNodeOperatorStakingLimit(_id, _stakingLimit);
     }
 
     /**
-      * @notice Report `_stoppedIncrement` more stopped validators of the node operator #`_id`
-      */
-    function reportStoppedValidators(uint256 _id, uint64 _stoppedIncrement) external
+     * @notice Report `_stoppedIncrement` more stopped validators of the node operator #`_id`
+     */
+    function reportStoppedValidators(uint256 _id, uint64 _stoppedIncrement)
+        external
         authP(REPORT_STOPPED_VALIDATORS_ROLE, arr(_id, uint256(_stoppedIncrement)))
         operatorExists(_id)
     {
         require(0 != _stoppedIncrement, "EMPTY_VALUE");
-        operators[_id].stoppedValidators = operators[_id].stoppedValidators.add(_stoppedIncrement);
-        require(operators[_id].stoppedValidators <= operators[_id].usedSigningKeys, "STOPPED_MORE_THAN_LAUNCHED");
 
-        emit NodeOperatorTotalStoppedValidatorsReported(_id, operators[_id].stoppedValidators);
+        uint64 newStoppedValidatorsCount = operators[_id].stoppedValidators.add(_stoppedIncrement);
+        require(newStoppedValidatorsCount <= operators[_id].usedSigningKeys, "STOPPED_MORE_THAN_LAUNCHED");
+
+        operators[_id].stoppedValidators = newStoppedValidatorsCount;
+
+        uint64 currentStakingLimit = operators[_id].stakingLimit;
+        if (currentStakingLimit > 0) {
+            uint64 stakingLimitDecrease = currentStakingLimit > _stoppedIncrement ? _stoppedIncrement : currentStakingLimit;
+            _setNodeOperatorStakingLimit(_id, operators[_id].stakingLimit.sub(stakingLimitDecrease));
```

Comment:
Let's refactor all key logic alltogether
<img width="1137" alt="image" src="https://user-images.githubusercontent.com/1270671/209642530-c14a68df-b84d-436b-abc3-f897c2b59d54.png">

Reply:
![Untitled(1)](https://user-images.githubusercontent.com/961317/209652440-6493866c-3315-47b0-8cb8-832fbdb11126.jpg)

Reply:
![StakingRouter conventions about keys and validators](https://user-images.githubusercontent.com/961317/209662588-ff2ec083-a097-403d-a41c-3fa03205b12c.jpg)